calico_node/allocateipip/allocate_ipip_addr.go

```diff
@@ -35,68 +42,86 @@ func main() {
+	}
+
+	if node.Spec.BGP.IPv4IPIPTunnelAddr == "" {
 		// The IPIP tunnel has no IP address assigned, assign one.
 		log.Debug("IPIP tunnel is not assigned - assign IP")
-		assignHostTunnelAddr(c, nodename, ipipCidrs)
-	} else if isIpInPool(ipAddr, ipipCidrs) {
+		assignHostTunnelAddr(ctx, c, nodename, ipipCidrs)
+	} else if isIpInPool(node.Spec.BGP.IPv4IPIPTunnelAddr, ipipCidrs) {
 		// The IPIP tunnel address is still valid, so leave as it.
-		log.WithField("IP", ipAddr.String()).Info("IPIP tunnel address is still valid")
+		log.WithField("IP", node.Spec.BGP.IPv4IPIPTunnelAddr).Info("IPIP tunnel address is still valid")
 	} else {
 		// The address that is currently assigned is no longer part
 		// of an IPIP pool, so release the IP, and reassign.
-		log.WithField("IP", ipAddr.String()).Info("Reassigning IPIP tunnel address")
+		log.WithField("IP", node.Spec.BGP.IPv4IPIPTunnelAddr).Info("Reassigning IPIP tunnel address")
+		ipAddr := net.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to parse the CIDR '%s'", node.Spec.BGP.IPv4IPIPTunnelAddr)
+		}
 
 		ipsToRelease := []net.IP{*ipAddr}
-		_, err := c.IPAM().ReleaseIPs(ipsToRelease)
+		_, err := c.IPAM().ReleaseIPs(ctx, ipsToRelease)
 		if err != nil {
 			log.WithField("IP", ipAddr.String()).WithError(err).Fatal("Error releasing non IPIP address")
 		}
 
 		// Assign a new tunnel address.
-		assignHostTunnelAddr(c, nodename, ipipCidrs)
+		assignHostTunnelAddr(ctx, c, nodename, ipipCidrs)
 	}
 }
 
 // removeHostTunnelAddr removes any existing IP address for this host's IPIP
 // tunnel device and releases the IP from IPAM.  If no IP is assigned this function
 // is a no-op.
-func removeHostTunnelAddr(c *client.Client, nodename string) {
-	if ipAddr, err := c.Config().GetNodeIPIPTunnelAddress(nodename); err != nil {
-		log.WithError(err).Fatal("Unable to retrieve IPIP tunnel address for cleanup")
-	} else if ipAddr == nil {
+func removeHostTunnelAddr(ctx context.Context, c client.Interface, nodename string) {
+	node, err := c.Nodes().Get(ctx, nodename, options.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatalf("Unable to retrieve IPIP tunnel address for cleanup. Error getting node '%s'", nodename)
+	}
+
+	if node.Spec.BGP.IPv4IPIPTunnelAddr == "" {
 		log.Debug("No IPIP tunnel address assigned, and not required")
-	} else if _, err := c.IPAM().ReleaseIPs([]net.IP{*ipAddr}); err != nil {
+		return
+	}
+
+	ipAddr := net.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+	if _, err := c.IPAM().ReleaseIPs(ctx, []net.IP{*ipAddr}); err != nil {
 		log.WithError(err).WithField("IP", ipAddr.String()).Fatal("Error releasing IPIP address from IPAM")
-	} else if err = c.Config().SetNodeIPIPTunnelAddress(nodename, nil); err != nil {
+	}
+
+	node.Spec.BGP.IPv4IPIPTunnelAddr = ""
+	if _, err = c.Nodes().Update(ctx, node, options.SetOptions{}); err != nil {
```

Comment:
Feels like this should also be encased in a retry loop to handle the conflict error scenario.